addons/arsenal/functions/fnc_attributeAddCompatible.sqf

```diff
@@ -0,0 +1,75 @@
+#include "script_component.hpp"
+#include "..\defines.hpp"
+/*
+ * Author: mharis001
+ * Adds compatible attachments or magazines for all weapons in 3DEN attribute.
+ *
+ * Arguments:
+ * 0: Attribute controls group <CONTROL>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [CONTROL] call ace_arsenal_fnc_attributeAddCompatible
+ *
+ * Public: No
+ */
+
+params ["_controlsGroup"];
+
+private _category = lbCurSel (_controlsGroup controlsGroupCtrl IDC_ATTRIBUTE_CATEGORY);
+
+// Exit if selected category is not attachments or magazines
+if !(_category in [4, 5, 6, 7, 8]) exitWith {};
+
+private _configItems = +(uiNamespace getVariable [QGVAR(configItems), []]);
+private _attributeValue = uiNamespace getVariable [QGVAR(attributeValue), [[], 0]];
+_attributeValue params ["_attributeItems"];
+
+// Get list of all weapons in attribute items
+(_configItems select 0) params ["_primaryWeapons", "_secondaryWeapons", "_handgunWeapons"];
+private _attributeWeapons = _attributeItems select {_x in _primaryWeapons || {_x in _secondaryWeapons} || {_x in _handgunWeapons}};
+
+// Add compatible attachments or magazines to attribute
+private _cfgWeapons = configFile >> "CfgWeapons";
+private _itemsToAdd = [];
+
+if (_category == 8) then {
+    private _magazineGroups = +(uiNamespace getVariable QGVAR(magazineGroups));
+    private _cfgMagazines = configFile >> "CfgMagazines";
+
+    {
+        private _weaponConfig = _cfgWeapons >> _x;
+
+        {
+            private _muzzleConfig = if (_x == "this") then {_weaponConfig} else {_weaponConfig >> _x};
+
+            // Only add existent magazines and ensure correct classname case
+            private _magazines = getArray (_muzzleConfig >> "magazines") select {isClass (_cfgMagazines >> _x)};
```

Comment:
CBA_fnc_compatibleMagazines?

Reply:
We have to. For magazineWells.

Reply:
Can change to avoid duplicate code. Will need to test the performance difference for a large weapons list.

Reply:
I guess this does match code in fillRightPanel
cba func does cache results, but on first call it probably will be more expensive

Reply:
Did a quick test with ~400 weapons (RHS) and `diag_tickTime`, here are the results for the first and second clicks.
Current:
```
14:08:41 [ACE] (arsenal) INFO: Added compatible magazines (current) - Time: 36.9873 ms
14:09:05 [ACE] (arsenal) INFO: Added compatible magazines (current) - Time: 37.9944 ms
```
CBA_fnc_compatibleMagazines:
```
14:10:59 [ACE] (arsenal) INFO: Added compatible magazines (cba func) - Time: 67.0166 ms
14:11:13 [ACE] (arsenal) INFO: Added compatible magazines (cba func) - Time: 14.0076 ms
```
So not too big of a difference and this is also the extreme case of all weapons. Thoughts?

Reply:
That only happens once when you open the UI right?
Perf of both is fine for that.
CBA func uses additional memory for it's cache. But reduces code duplication.
So don't really know.
CBA func also filters out non existent magazines (which can occur in magazinewells) which you don't really do (configName will probably return "" so you'll have a bunch of empty strings)

Reply:
The arsenal magazine wells cache also filters out non-existent magazines aswell: https://github.com/acemod/ACE3/blob/master/addons/arsenal/functions/fnc_scanConfig.sqf#L180

Yeah, the performance is fine for both (and I can't see the button being clicked very often). The slowest part is L72 which in the above situation takes 900-1000 ms.

Reply:
Don't really need to worry about 35ms if you have a 1000ms laying around a couple lines lower.

It really doesn't matter. Whatever you prefer. Less code duplication is the only reason that I have left as a reason to change this.

Reply:
Yeah, I think in that case it is fine as is.